File: cpp/validPath.cpp
#include <vector>
#include <map>
#include <queue>
#include <unordered_set>

using namespace std;

class Solution {
public:
    bool validPath(int n, vector<vector<int>>& edges, int source, int destination) {
        map<int, vector<int>> mp;
        unordered_set<int> v;
        

        for (vector<int> v :edges) {
            mp[v[0]].push_back(v[1]);
            mp[v[1]].push_back(v[0]);
        }

        queue<int> q;
        q.push(source);
        v.insert(source);

        while (!q.empty()) {
            int c = q.front();
            q.pop();
            if (c == destination) return true;

            for (int k : mp[c]) {
                if (v.find(k) == v.end()) {
                    v.insert(k);
                    q.push(k);
                }
            }
        }
        return false;
    }
};

File: validPath.py
from collections import deque

class Solution:
    def validPath(self, n: int, edges: list[list[int]], source: int, destination: int) -> bool:
        mp:dict = {}
        for i, j in edges:
            if i not in mp.keys():
                mp[i] = []
            if j not in mp.keys():
                mp[j] = []
            mp[i].append(j)
            mp[j].append(i)
        
        q = deque()
        v = set()
        v.add(source)
        q.append(source)

        while q:
            c = q.popleft()
            if c == destination:
                return True
            
            for k in mp[c]:
                if k not in v:
                    q.append(k)
                    v.add(k)
        return False